### ISSUE TITLE
problems with --modify-files

### ISSUE THREAD

**Issue by finanalyst** [lizmat/App-Rak#52]
***rak*** with `--modify-files` produces (I think) incorrect reports and does not allow non-destructive potential results.
a) A zipped directory is attached with three files in a directory 'rak-test'
[raktest.zip](https://github.com/lizmat/App-Rak/files/14613987/raktest.zip)
b) Two files contain 'processed', one contains 'xrocessed', each file has 9 lines
c) The following rak commands and output were obtained
```
rak '*.subst(/ "processed" /, ":processing")' rak-test/ --modify-files --dryrun
Processed 3 files, 3 files changed, 27 lines changed
*** no changes where made because of --dryrun ***
$ rak '/ "processed" /' rak-test/ --files-with-matches
rak-test/01.rakutest
rak-test/02.rakutest
```
After running the first stanza, I would have expected the result to be `3 files, 2 files changed, 4 lines changed`

Further, I wanted to see what might be changed.
```
$ rak '*.subst(/ "processed" /, ":processing")' rak-test/ --modify-files --dryrun --files-with-matches
The --files-with-matches option is not being handled and will be ignored.
If you believe this to be incorrect, please report an issue with <snip>
```
I have had problems with `--modify-files` because I have mis-specified the regex and changed too much or too many files with no way of easily determining what is being picked up.

Perhaps the output to `--dryrun --verbose` could be each line changed with the original followed by the changed line (coloured differently ?).

Info:
```
$ rak --version
rak - provided by App::Rak 0.2.20, running Raku 6.d with Rakudo 2024.02.
```

**Comment by lizmat** [comment]
Ok, it looks like it is not going to be an easy fix.  Please use the workaround for now.

**Comment by finanalyst** [comment]
If `.subst` returns a string whether it is changed or not, and a string coerces to True, then `.subst` will report all lines in all files will be changed, so under this view, the problem is not rak, but `.subst`
Suppose, for `rak` you introduce a new method `.on-subst` that returns a string - True - if the match succeeds on a string, and returns False if the match does not exist, or mixes in the Bool value of the match to the string, so that when the match succeeds, the string has a True value when coerced to a Bool, and a False value if the match does not succeed when when coerced to a Bool.

**Comment by lizmat** [comment]
The semantics of `.subst` are by now more or less cast in stone.

It's the `rak` module that needs fixing: it needs to check whether the returned string is the same as the source string, and *not* mark it as changed.  As per the documentation  :-)

**Comment by lizmat** [comment]
This has been fixed with release 0.3.1, now in he ecosystem!
